plugins/source/gcp/codegen/client/discover.go

```diff
@@ -0,0 +1,44 @@
+package client
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"sort"
+
+	"golang.org/x/exp/maps"
+)
+
+type Service struct {
+	Name string `json:"name"`
+}
+
+func Discover() ([]string, error) {
+	resp, err := http.Get("https://discovery.googleapis.com/discovery/v1/apis")
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	var services struct {
+		Items []Service `json:"items"`
+	}
+	err = json.Unmarshal(body, &services)
+	if err != nil {
+		return nil, err
+	}
+
+	servicesMap := make(map[string]bool)
+	for _, service := range services.Items {
+		servicesMap[service.Name+".googleapis.com"] = true
```

Comment:
This is require since there are multiple versions of the same services with the same name (e.g. v1, v1Beta). The API has a different ID for each version

Reply:
Instead of generating the domain name, why not just parse the DiscoveryURL and extract the hostname:

```
u := url.Parse(service.DiscoveryRestUrl)
hostname := u.Hostname()
servicesMap[hostname] = true
```

Reply:
I tried that but it doesn't work for some services (e.g. compute):
`"discoveryRestUrl": "https://www.googleapis.com/discovery/v1/apis/compute/alpha/rest"`

Reply:
Fair!